Source/GTMSessionFetcher.m

```diff
@@ -1689,8 +1704,72 @@ - (void)authorizer:(id<GTMFetcherAuthorizationProtocol>)auth
     @synchronized(self) {
       _request = authorizedRequest;
     }
-    [self beginFetchMayDelay:NO mayAuthorize:NO];
+    [self beginFetchMayDelay:NO mayAuthorize:NO mayDecorate:YES];
+  }
+}
+
+- (void)decorateHeadersForRequest:(NSURLRequest *)fetchRequest
+                 headerDecorators:(NSArray<id<GTMFetcherHeaderDecoratorProtocol>> *)headerDecorators {
+  GTMSessionCheckNotSynchronized(self);
+
+  GTMSESSION_LOG_DEBUG(@"GTMSessionFetcher decorateHeaders start %zu decorators", headerDecorators.count);
+
+  dispatch_queue_t decorateCompletionQueue;
```

Comment:
The authorization protocol provides the means to stop authorization, if necessary (`stopAuthorizationForRequest:`). Given the implementation here, it seems it would be difficult to implement a similar scheme in case a decorator takes too much time and a fetch is stopped, no?

Reply:
Yeah, that's the downside of using blocks. However, the worst thing that happens is the header calculation result is thrown away when the decorator invokes its callback, right?

IMHO, the benefits of blocks (type safety, etc.) outweigh the benefits of allowing early cancellation of calculating headers. What do you think?